xarray_beam/_src/pangeo_forge.py

```diff
@@ -0,0 +1,153 @@
+
+from xarray_beam._src import core, rechunk
+
+
+def _zero_dimensions(dataset: xarray.Dataset) -> Mapping[str, int]:
+  return {dim: 0 for dim in dataset.dims.keys()}
+
+
+def _expand_dimensions_by_key(
+    dataset: xarray.Dataset,
+    index: 'FilePatternIndex',
+    pattern: 'FilePattern'
+) -> xarray.Dataset:
+  """Expand the dimensions of the `Dataset` by offsets found in the `Key`."""
+  combine_dims_by_name = {
+    combine_dim.name: combine_dim for combine_dim in pattern.combine_dims
+  }
+  index_by_name = {
+    idx.name: idx for idx in index
+  }
+
+  if not combine_dims_by_name:
+    return dataset
+
+  for dim_key in index_by_name.keys():
+    # skip expanding dimensions if they already exist
+    if dim_key in dataset.dims:
+      continue
+
+    try:
+      combine_dim = combine_dims_by_name[dim_key]
+    except KeyError:
+      raise ValueError(
+        f"could not find CombineDim named {dim_key!r} in pattern {pattern!r}."
+      )
+
+    dim_val = combine_dim.keys[index_by_name[dim_key].index]
+    dataset = dataset.expand_dims(**{dim_key: [dim_val]})
+
+  return dataset
+
+
+class FilePatternToChunks(beam.PTransform):
+  """Open data described by a Pangeo-Forge `FilePattern` into keyed chunks."""
+
+  from pangeo_forge_recipes.patterns import FilePattern, FilePatternIndex
+
+  def __init__(
+      self,
+      pattern: 'FilePattern',
+      sub_chunks: Optional[Mapping[str, int]] = None,
+      xarray_open_kwargs: Optional[Dict] = None
+  ):
+    """Initialize FilePatternToChunks.
+
+    TODO(#29): Currently, `MergeDim`s are not supported.
+
+    Args:
+      pattern: a `FilePattern` describing a dataset.
+      sub_chunks: split each open dataset into smaller chunks. If not set, the
+        transform will return one file per chunk.
+      xarray_open_kwargs: keyword arguments to pass to `xarray.open_dataset()`.
+    """
+    self.pattern = pattern
+    self.sub_chunks = sub_chunks
+    self.xarray_open_kwargs = xarray_open_kwargs or {}
+    self._max_size_idx = {}
+
+    if pattern.merge_dims:
+      raise ValueError("patterns with `MergeDim`s are not supported.")
+
+  @contextlib.contextmanager
+  def _open_dataset(self, path: str) -> xarray.Dataset:
+    """Open as an XArray Dataset, sometimes with local caching."""
+    fs_file = None
+    with FileSystems().open(path) as file:
+      try:
+        dataset = xarray.open_dataset(file, **self.xarray_open_kwargs)
+      except (TypeError, OSError):
+        # The cfgrib engine (and others) may fail with the FileSystems method of
+        # opening with BufferedReaders. Here, we open the data locally to make
+        # it easier to work with XArray.
+        fs_file = fsspec.open_local(f"simplecache::{path}",
+                                    simplecache={'cache_storage': '/tmp/files'})
+        dataset = xarray.open_dataset(fs_file, **self.xarray_open_kwargs)
+
+      yield dataset
+
+      if fs_file:
+        fs_file.close()
```

Comment:
This  clean-up under `if fs_file` would ideally happen even if an exception is raised from calling `open_local()`.

I think a slightly cleaner pattern would be:
```suggestion
    with FileSystems().open(path) as file:
      try:
        yield xarray.open_dataset(file, **self.xarray_open_kwargs)
      except (TypeError, OSError):
        # The cfgrib engine (and others) may fail with the FileSystems method of
        # opening with BufferedReaders. Here, we open the data locally to make
        # it easier to work with XArray.
        with fsspec.open_local(f"simplecache::{path}",
                                simplecache={'cache_storage': '/tmp/files'}) as fs_file:
          yield xarray.open_dataset(fs_file, **self.xarray_open_kwargs)
```

But actually -- perhaps this could be an option like `local_copy=True` in `FilePatternToChunks` instead of using the fall-back? Otherwise, users might be surprised that local copies are being made (e.g., using extra disk space) that they thought were unnecessary.